File: solutions/typescript/052-search-a-2d-matrix.ts
/**
 * Problem 52: Search a 2D Matrix (LeetCode 74)
 * Difficulty: Med
 * Language: TypeScript
 */
function searchMatrix(matrix: number[][], target: number): boolean {
  const flat = matrix.flat();
  const go = (lo: number, hi: number): boolean => {
    if (lo > hi) return false;
    const mid = Math.floor((lo + hi) / 2);
    if (flat[mid] === target) return true;
    return flat[mid] < target ? go(mid + 1, hi) : go(lo, mid - 1);
  };
  return go(0, flat.length - 1);
}

File: solutions/ocaml/052-search-a-2d-matrix.ml
(*
  Problem 52: Search a 2D Matrix (LeetCode 74)
  Difficulty: Med
  Language: OCaml
*)
let search_matrix matrix target =
  let flat = List.concat matrix |> Array.of_list in
  let rec go lo hi =
    if lo > hi then false
    else
      let mid = (lo + hi) / 2 in
      if flat.(mid) = target then true
      else if flat.(mid) < target then go (mid + 1) hi
      else go lo (mid - 1)
  in go 0 (Array.length flat - 1)
